Add drag zoom feature with CTRL+MiddleMouseButton

zmn-hamid opened the pull request:
Zoom drag feature was only available in 3D editor (viewport). In this commit I added the same functionality to other editors as well, so you can zoom drag in any of them: 2D editor, animation track editor, visual shader editor, etc.

This keeps Godot consistent and immensely helps with 2D game development. My mouse wheel doesn't work well, I needed this change so much. And since I use it all the time in other software like Photoshop, I made this PR. You can pan using middle click and zoom by adding CTRL to the combo.

Bellow is a demo for the 2D editor:

https://github.com/user-attachments/assets/e3cc8076-2de9-4c54-a48d-44bfa2328d2a

zmn-hamid commented:
Fixed a slight issue where the inputs could be locked. For example If you hold MMB, then hold CTRL, then only release MMB, It would've been stuck with the dragging and wouldn't reset unless manually reset using the same key combination or LMB. Now it will reset when you let go of MMB. The opposite would still happen if you let go of CTRL (will be locked on zoom dragging), but once you let go of MMB, that will reset as well so no bad user experience will be felt.

zmn-hamid commented:
Fixed the PascalCase typo in commit message

Calinou reviewed (approved):
Tested locally on Linux, it works as expected in both the 2D editor and visual shader editor. Code looks good to me.

groud commented on scene/gui/view_panner.cpp:
```diff
@@ -92,19 +92,34 @@ bool ViewPanner::gui_input(const Ref<InputEvent> &p_event, Rect2 p_canvas_rect)
 			return false;
 		}
 
+		bool is_zoom_drag_event = mb->get_button_index() == MouseButton::MIDDLE && mb->is_ctrl_pressed();
```
I think this implementation makes the code a bit complicated and hard to read. I suggest using an enum with a DRAG_TYPE_ instead. You can check how it's done in the TileMap editors, but basically, the idea would to use something like this:
```
	enum DragType {
		DRAG_TYPE_NONE,
		DRAG_TYPE_PAN,
		DRAG_TYPE_ZOOM,
	};
```
This makes sure all states are mutually exclusive and there's no way the both zoom and pan are possible at the same time.

KoBeWi commented:
The view moves a lot while zooming. It's taking the current mouse position into account and tries to zoom on that position. It should always use the initial position.

zmn-hamid commented:
All the requests were reasonable and I added them. I myself didn't know it would work in graph and visual editors! Therefore I changed the commit message and the PR's title to match the true meaning of this commit. Hope it helps

KoBeWi commented:
>  I myself didn't know it would work in graph and visual editors!

It works in every editor that uses ViewPanner and there is a few of them.

The issue I mentioned before still exists in TileMap editor:

https://github.com/user-attachments/assets/e55e3070-3a7d-4865-adf6-b2a0286aa5a4

It does not happen in 2D and Polygon editors. Not sure what causes this difference.

zmn-hamid commented:
> > I myself didn't know it would work in graph and visual editors!
> 
> It works in every editor that uses ViewPanner and there is a few of them.
> 
> The issue I mentioned before still exists in TileMap editor:
> 
>  godot.windows.editor.dev.x86_64_GAhtRzMw6U.mp4 
> It does not happen in 2D and Polygon editors. Not sure what causes this difference.

Oh it's interesting. I'll look that up

zmn-hamid commented:
> The issue I mentioned before still exists in TileMap editor:

I've been trying to fix this bug since, and got no results. I'm new to cpp & software dev so maybe I don't understand. If anyone could pin point what needs to be changed that would be awesome. It should be related to `\editor\plugins\tiles\tile_data_editors.cpp` but I couldn't fix it at all. And it's weird because why is it even happening in tile set editor if it's not using `view_panner` at all? Maybe it's a bug?

KoBeWi commented:
It's in `tile_atlas_view.cpp`.

zmn-hamid commented:
Thanks for the help and suggestions. I fixed them in this patch

passivestar commented:
If it's added for consistency with 3D should it also have the same options that 3D has? I expect people who use those settings might want same behavior in both 2D and 3D

<img width="589" alt="image" src="https://github.com/user-attachments/assets/d1b583c1-dc96-4c56-8f6e-a30bc5bcfebc" />

zmn-hamid commented:
> If it's added for consistency with 3D should it also have the same options that 3D has? I expect people who use those settings might want same behavior in both 2D and 3D
> 
> <img width="589" alt="image" src="https://github.com/user-attachments/assets/d1b583c1-dc96-4c56-8f6e-a30bc5bcfebc" />
> 

Makes sense, I'll work on them. but I think we should instead separate these settings and make them more global. Because:
1. Less redundancy
2. The feature that I'm adding right now isn't limited to the 2D scene, it works on any editor. But the other one is only for the 3D viewport.
However I'm not sure under which category I should put in these new settings. Any suggestions?

passivestar commented:
> However I'm not sure under which category I should put in these new settings. Any suggestions?

I'm not sure tbh, could be `editors/common`. It seems like overall categories could use some reorganization, because while most sections in "Editors" could be described as "editors", there's also "Panning" which is not an editor, it would probably make more sense in `editors/common/panning` and then common zoom settings could be in `editors/common/zooming`. Also `editors/3d_gizmos` could probably be in `editors/3d/3d_gizmos` because "3D Gizmos" is not an editor either

There are some other things in settings overall that don't seem to make sense, like `interface/editor` and `interface/editors` being separate categories.

Need to hear from maintainers on what to do about it and whether there are any plans to clean up the categories (maybe a lot at once in a single engine update so that it's easier to migrate them), then it will be easier to decide where to add these.

zmn-hamid commented:
Actually as I'm thinking about it, I should put it in `editors/panning/` for now because this commit is really not about 2D scene anymore

zmn-hamid commented:
I only added the `zoom_style` in this patch because in order to add the `zoom_mouse_button` I had to register a modifier as well, like in `editor\plugins\node_3d_editor_plugin.cpp:5715` but that file is only for 3d scene. I prefer not messing anymore and let this be added once the settings have been reorganized.

zmn-hamid commented:
Fixed a small initialization issue.

zmn-hamid committed:
Add drag zoom feature with CTRL+MiddleMouseButton

This change introduces a DragType enum to scene/gui/view_panner.cpp of
dragging, which includes:

- DRAG_TYPE_NONE: Not dragging
- DRAG_TYPE_PAN: Panning (dragging using MMB)
- DRAG_TYPE_ZOOM: Zooming (dragging using CTRL+MMB)

The goal of this change is the third option, which was already available
in 3D viewport but not in 2D. This feature should work in other editors
as well such as Animation Track Editor and Visual Shader Editor and so
on.

zmn-hamid commented:
Accidentally removed two lines from the code. Added them back.

Calinou reviewed (approved):
Tested again, it works as expected. Code looks good to me.

I think the settings are good as they are. We can revisit this later if there's a need to add more settings.

zmn-hamid commented:
Thanks for the suggestions by @AThousandShips. Refactored the code.

zmn-hamid commented on editor/plugins/tiles/tile_atlas_view.h:
```diff
@@ -58,7 +58,7 @@ class TileAtlasView : public Control {
 	CenterContainer *center_container = nullptr;
 	Vector2 panning;
-	void _update_zoom_and_panning(bool p_zoom_on_mouse_pos = false);
+	void _update_zoom_and_panning(bool p_zoom_on_mouse_pos = false, Vector2 p_mouse_pos = Vector2());
```
This resulted in this error:
```
editor/plugins/tiles/tile_atlas_view.cpp:95:21: error: out-of-line definition of '_update_zoom_and_panning' does not match any declaration in 'TileAtlasView'
```
Should I add the same declaration to the cpp file?

AThousandShips replied:
Yes, they should always match (didn't add the suggestion because I thought it was redundant)

zmn-hamid replied:
> Yes, they should always match (didn't add the suggestion because I thought it was redundant)

Oh ok thanks. I'm new to cpp, I thought that's intended

zmn-hamid commented:
Fixed a small bug in refactoring

AThousandShips commented on editor/plugins/tiles/tile_atlas_view.cpp:
```diff
@@ -132,8 +132,13 @@ void TileAtlasView::_update_zoom_and_panning(bool p_zoom_on_mouse_pos) {
 	if (p_zoom_on_mouse_pos) {
-		// Offset the panning relative to the center of panel.
-		Vector2 relative_mpos = get_local_mouse_position() - get_size() / 2;
+		Vector2 relative_mpos = Vector2();
```
```suggestion
		Vector2 relative_mpos;
```
Unnecessary default

zmn-hamid replied:
Are you sure? Because I remember as I was building the app it throwed an initialization error. You can see earlier I added a commit about it. 

Btw the code has been changed a bit now. Thanks for the review

zmn-hamid replied:
Oh sorry you're right. It was something else.

zmn-hamid commented:
It should be fixed now.

zmn-hamid commented:
Do I need to do anything RN?

groud commented on editor/plugins/tiles/tile_atlas_view.cpp:
```diff
@@ -132,8 +132,12 @@ void TileAtlasView::_update_zoom_and_panning(bool p_zoom_on_mouse_pos) {
-		// Offset the panning relative to the center of panel.
-		Vector2 relative_mpos = get_local_mouse_position() - get_size() / 2;
+		Vector2 relative_mpos;
+		if (p_mouse_pos == Vector2()) {
```
This does not look like the right way to do it. Using `== Vector2()` as a boolean value might break things if your mouse is at `(0,0)`. If you want to add another argument there it should not be optional I feel. Or, at least, if `p_zoom_on_mouse_pos` is true, then `p_mouse_pos` should be expected to be provided too.

KoBeWi replied:
You can use `Vector2(Mat::INF, Mat::INF)` for optional value and `Vector2.is_inf()` to check it.

zmn-hamid replied:
> You can use `Vector2(Mat::INF, Mat::INF)` for optional value and `Vector2.is_inf()` to check it.

Oh thanks for the suggestion! didn't know that.

groud replied:
In general I would avoid that though, it's a bit hacky. Especially with float numbers I think.
I think both arguments should be mandatory, and `p_mouse_pos` expected to be correct in case you use `p_zoom_on_mouse_pos`

zmn-hamid commented:
Updated the `_update_zoom_and_panning` to require `p_mouse_pos` when `p_zoom_on_mouse_pos` is `true`. It will be provided by `p_origin` in `TileAtlasView::_zoom_callback`. I over-complicated it in the previous commits
